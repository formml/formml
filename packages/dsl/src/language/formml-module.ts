import {
  createDefaultModule,
  createDefaultSharedModule,
  DefaultSharedModuleContext,
  inject,
  LangiumServices,
  LangiumSharedServices,
  Module,
  PartialLangiumServices,
} from 'langium'
import {
  FormMLGeneratedModule,
  FormMLGeneratedSharedModule,
} from './generated/module'
import { FormMLValidator, registerValidationChecks } from './formml-validator'

/**
 * Declaration of custom services - add your own service classes here.
 */
export type FormMLAddedServices = {
  validation: {
    FormMLValidator: FormMLValidator
  }
}

/**
 * Union of Langium default services and your custom services - use this as constructor parameter
 * of custom service classes.
 */
export type FormMLServices = LangiumServices & FormMLAddedServices

/**
 * Dependency injection module that overrides Langium default services and contributes the
 * declared custom services. The Langium defaults can be partially specified to override only
 * selected services, while the custom services must be fully specified.
 */
export const FormMLModule: Module<
  FormMLServices,
  PartialLangiumServices & FormMLAddedServices
> = {
  validation: {
    FormMLValidator: () => new FormMLValidator(),
  },
}

/**
 * Create the full set of services required by Langium.
 *
 * First inject the shared services by merging two modules:
 *  - Langium default shared services
 *  - Services generated by langium-cli
 *
 * Then inject the language-specific services by merging three modules:
 *  - Langium default language-specific services
 *  - Services generated by langium-cli
 *  - Services specified in this file
 *
 * @param context Optional module context with the LSP connection
 * @returns An object wrapping the shared services and the language-specific services
 */
export function createFormMLServices(context: DefaultSharedModuleContext): {
  shared: LangiumSharedServices
  FormML: FormMLServices
} {
  const shared = inject(
    createDefaultSharedModule(context),
    FormMLGeneratedSharedModule
  )
  const FormML = inject(
    createDefaultModule({ shared }),
    FormMLGeneratedModule,
    FormMLModule
  )
  shared.ServiceRegistry.register(FormML)
  registerValidationChecks(FormML)
  return { shared, FormML }
}
