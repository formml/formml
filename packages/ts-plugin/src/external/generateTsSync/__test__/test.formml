form testForm {
  num  numField
  text textField
}
