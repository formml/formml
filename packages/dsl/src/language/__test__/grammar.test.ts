import { parseHelper } from 'langium/test'

import { createInMemoryServices } from '../formml-module.js'
import { FormMLSchema } from '../generated/ast.js'

describe('grammar', () => {
  const services = createInMemoryServices()
  const serialize = (ast: FormMLSchema) =>
    services.FormML.serializer.JsonSerializer.serialize(ast, { space: 2 })
  const parser = (input: string) =>
    parseHelper<FormMLSchema>(services.FormML)(input)
      .then((x) => x.parseResult)
      .then((r) =>
        r.lexerErrors.length > 0 || r.parserErrors.length > 0
          ? Promise.reject(new Error('Parsing failed'))
          : r.value,
      )

  describe('syntax', () => {
    test('comments', async () => {
      const content = `
        // inline comment

        form Example {
          /* multiple line comment
          num commentedField
          */

          num numberField
        }
      `

      const ast = await parser(content)
      expect(serialize(ast)).toMatchSnapshot()
    })
  })

  describe('simple fields', () => {
    test('primitives', async () => {
      const content = `
        form ExampleForm {
          num      numberField
          decimal  decimalField
          text     textField
          bool	   boolField
          datetime datetimeField
        }
      `
      const ast = await parser(content)
      expect(serialize(ast)).toMatchSnapshot()
    })

    test('invalidate unknown types', async () => {
      const content = `
        form ExampleForm {
          unknown invalidType
        }
      `
      await expect(() => parser(content)).rejects.toThrow()
    })
  })

  describe('validation annotations', () => {
    describe('non-argument annotation', () => {
      test('one line', async () => {
        const content = `
          form ExampleForm {
            @required num numberField
          }
        `
        const ast = await parser(content)
        expect(serialize(ast)).toMatchSnapshot()
      })

      test('multiple lines', async () => {
        const content = `
          form ExampleForm {
            @required
            num numberField
          }
        `
        const ast = await parser(content)
        expect(serialize(ast)).toMatchSnapshot()
      })

      test('multiple annotations', async () => {
        const content = `
          form ExampleForm {
            @required1
            @required2
            num numberField
          }
        `
        const ast = await parser(content)
        expect(serialize(ast)).toMatchSnapshot()
      })

      test('annotation with parentheses', async () => {
        const content = `
          form ExampleForm {
            @required()
            num numberField
          }
        `
        const ast = await parser(content)
        expect(serialize(ast)).toMatchSnapshot()
      })
    })

    describe('annotation with arguments', () => {
      test('one argument', async () => {
        const content = `
          form ExampleForm {
            @min(10)
            num numberField
          }
        `
        const ast = await parser(content)
        expect(serialize(ast)).toMatchSnapshot()
      })

      describe('argument type', () => {
        describe('JS-like literals', () => {
          test.each([`'single quotes'`, `"double quotes"`])(
            'string - %j',
            async (input) => {
              const content = `
                form ExampleForm {
                  @any(${input})
                  num numberField
                }
              `
              const ast = await parser(content)
              expect(serialize(ast)).toMatchSnapshot()
            },
          )

          test.each(['0', '123', '-123', '123.456', '-123.456'])(
            'number - %j',
            async (input) => {
              const content = `
                form ExampleForm {
                  @any(${input})
                  num numberField
                }
              `
              const ast = await parser(content)
              expect(serialize(ast)).toMatchSnapshot()
            },
          )
        })
      })
    })
  })
})
