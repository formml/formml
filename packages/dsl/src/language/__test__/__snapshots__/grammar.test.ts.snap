// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`grammar > simple fields > primitives 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "ExampleForm",
    "fields": [
      {
        "$type": "Field",
        "type": "num",
        "name": "numberField",
        "annotations": []
      },
      {
        "$type": "Field",
        "type": "decimal",
        "name": "decimalField",
        "annotations": []
      },
      {
        "$type": "Field",
        "type": "text",
        "name": "textField",
        "annotations": []
      },
      {
        "$type": "Field",
        "type": "bool",
        "name": "boolField",
        "annotations": []
      },
      {
        "$type": "Field",
        "type": "datetime",
        "name": "datetimeField",
        "annotations": []
      }
    ]
  }
}"
`;

exports[`grammar > syntax > comments 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "Example",
    "fields": [
      {
        "$type": "Field",
        "type": "num",
        "name": "numberField",
        "annotations": []
      }
    ]
  }
}"
`;

exports[`grammar > validation annotations > annotation with arguments > argument type > JS-like literals > string - "'single quotes'" 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "ExampleForm",
    "fields": [
      {
        "$type": "Field",
        "annotations": [
          {
            "$type": "Annotation",
            "name": "any",
            "args": [
              {
                "$type": "SQString",
                "value": "'single quotes'"
              }
            ]
          }
        ],
        "type": "num",
        "name": "numberField"
      }
    ]
  }
}"
`;

exports[`grammar > validation annotations > annotation with arguments > argument type > JS-like literals > string - "\\"double quotes\\"" 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "ExampleForm",
    "fields": [
      {
        "$type": "Field",
        "annotations": [
          {
            "$type": "Annotation",
            "name": "any",
            "args": [
              {
                "$type": "DQString",
                "value": "\\"double quotes\\""
              }
            ]
          }
        ],
        "type": "num",
        "name": "numberField"
      }
    ]
  }
}"
`;

exports[`grammar > validation annotations > annotation with arguments > one argument 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "ExampleForm",
    "fields": [
      {
        "$type": "Field",
        "annotations": [
          {
            "$type": "Annotation",
            "name": "min",
            "args": [
              {
                "$type": "Number_",
                "value": 10
              }
            ]
          }
        ],
        "type": "num",
        "name": "numberField"
      }
    ]
  }
}"
`;

exports[`grammar > validation annotations > non-argument annotation > annotation with parentheses 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "ExampleForm",
    "fields": [
      {
        "$type": "Field",
        "annotations": [
          {
            "$type": "Annotation",
            "name": "required",
            "args": []
          }
        ],
        "type": "num",
        "name": "numberField"
      }
    ]
  }
}"
`;

exports[`grammar > validation annotations > non-argument annotation > multiple annotations 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "ExampleForm",
    "fields": [
      {
        "$type": "Field",
        "annotations": [
          {
            "$type": "Annotation",
            "name": "required1",
            "args": []
          },
          {
            "$type": "Annotation",
            "name": "required2",
            "args": []
          }
        ],
        "type": "num",
        "name": "numberField"
      }
    ]
  }
}"
`;

exports[`grammar > validation annotations > non-argument annotation > multiple lines 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "ExampleForm",
    "fields": [
      {
        "$type": "Field",
        "annotations": [
          {
            "$type": "Annotation",
            "name": "required",
            "args": []
          }
        ],
        "type": "num",
        "name": "numberField"
      }
    ]
  }
}"
`;

exports[`grammar > validation annotations > non-argument annotation > one line 1`] = `
"{
  "$type": "FormMLSchema",
  "form": {
    "$type": "Form",
    "name": "ExampleForm",
    "fields": [
      {
        "$type": "Field",
        "annotations": [
          {
            "$type": "Annotation",
            "name": "required",
            "args": []
          }
        ],
        "type": "num",
        "name": "numberField"
      }
    ]
  }
}"
`;
